test/parallel/test-stream-readable-unshift.js

```diff
@@ -0,0 +1,187 @@
+    string
+  );
+
+}
+
+{
+  // Check that data goes at the beginning
+  const readable = new Readable({ read() {} });
+  const string = 'front';
+
+  readable.push('b');
+  readable.unshift(string);
+
+  assert.strictEqual(
+    readable._readableState.buffer.head.data.toString('utf8'),
+    string
+  );
+}
+
+{
+  // Check that buffer is saved with correct encoding
+  const readable = new Readable({ read() {} });
+
+  const encoding = 'base64';
+  const string = Buffer.from('abc').toString(encoding);
+  readable.unshift(string, encoding);
+
+  assert.strictEqual(
+    readable._readableState.buffer.head.data.toString(encoding),
+    string
+  );
+}
+
+{
+
+  const streamEncoding = 'base64';
+
+  function checkEncoding(readable) {
+
+    // chunk encodings
+    const encodings = ['utf8', 'binary', 'hex', 'base64'];
+    for (const encoding of encodings) {
+      const string = 'abc';
+      readable.unshift(string, encoding);
+
+      // If encoding is the same as the state.encoding the string is
+      // saved as is
+      const expect = encoding !== streamEncoding ?
+        Buffer.from(string, encoding).toString(streamEncoding) : string;
+
+      // When using .setEncoding, check that the data is saved in the bufferList
+      // as a string & also that is saved in the encoding set in .setEncoding
+      assert.strictEqual(
+        readable._readableState.buffer.head.data,
+        expect
+      );
+    }
+  }
+
+  const r1 = new Readable({ read() {} });
+  r1.setEncoding(streamEncoding);
+  checkEncoding(r1);
+
+  const r2 = new Readable({ read() {}, encoding: streamEncoding });
+  checkEncoding(r2);
+
+}
+
+{
+  // Both .push & .unshift should have the same behaviour
+  // When setting an encoding, each chunk should be emitted with that encoding
+  const encoding = 'base64';
+
+  function checkEncoding(readable) {
+    const string = 'abc';
+    readable.on('data', common.mustCall((chunk) => {
+      assert.strictEqual(chunk, Buffer.from(string).toString(encoding));
+    }, 2));
+
+    readable.push(string);
+    readable.unshift(string);
+  }
+
+  const r1 = new Readable({ read() {} });
+  r1.setEncoding(encoding);
+  checkEncoding(r1);
+
+  const r2 = new Readable({ read() {}, encoding });
+  checkEncoding(r2);
+
+}
+
+{
+  // Check that error is thrown for invalid chunks
+
+  const readable = new Readable({ read() {} });
+  function checkError(fn) {
+    common.expectsError(fn, {
+      code: 'ERR_INVALID_ARG_TYPE',
+      type: TypeError
+    });
+  }
+
+  checkError(() => readable.unshift([]));
+  checkError(() => readable.unshift({}));
+  checkError(() => readable.unshift(0));
+
+}
+
+{
+  // Check that ObjectMode works
+  const readable = new Readable({ objectMode: true, read() {} });
+
+  const chunks = ['a', 1, {}, []];
+
+  for (const chunk of chunks) {
+    readable.unshift(chunk);
+
+    assert.strictEqual(
+      readable._readableState.buffer.head.data,
+      chunk
+    );
```

Comment:
This test can be rewritten without using internals, it should consume the stream and check that the chunks comes out in order.